src/shared/components/case-editor/services/page-validation.service.ts

```diff
@@ -20,19 +21,22 @@ export class PageValidationService {
       });
   }
 
+  public isHidden(caseField: CaseField, editForm: FormGroup): boolean {
+    const formFields = editForm.getRawValue();
+    if (caseField.field_type.complex_fields && caseField.field_type.complex_fields.length) {
+      return caseField.field_type.complex_fields.some((cf: CaseField) => !this.isHidden(cf, editForm))
```

Comment:
Trying to understand this bit of logic now... but needs a semicolon on the end either way.
`return caseField.field_type.complex_fields.some((cf: CaseField) => !this.isHidden(cf, editForm));`

`complex_fields => [ shown_field, shown_field ]`
=> `.some(...) => true`, which says that `caseField` _is_ hidden, even though all of the fields it contains are shown?
`complex_fields => [ hidden_field, hidden_field_2 ]`
=> `.some(...) => false`, which says that `caseField` _isn't_ hidden, even though all of the fields it contains _are_ hidden...?

Maybe it should be this?
`!return caseField.field_type.complex_fields.some((cf: CaseField) => !this.isHidden(cf, editForm));`

`complex_fields => [ shown_field, shown_field ]`
=> `!.some(...) => false`, which says that `caseField` _isn't_ hidden, which is what we want.
`complex_fields => [ hidden_field, hidden_field_2 ]`
=> `!.some(...) => true`, which says that `caseField` _is_ hidden since none of the fields within it are shown, which is also what we want.

Or have I misunderstood the intent here? 😬

Reply:
@paul-graham The logic should be if we have some hidden fields do not bother checking the validity of the fields inside that complex.
`if (caseField.field_type.complex_fields && caseField.field_type.complex_fields.length) {
      const hiddenFields = caseField.field_type.complex_fields.filter((cf: CaseField) => this.isHidden(cf, editForm))
      return !hiddenFields.length
    }`

Reply:
But that still doesn't make any sense to me, which is why I'm struggling to understand what you're trying to do with this. This is what it appears to be doing (in pseudocode):

```
function isHidden(field: CaseField): boolean {
  if (caseField.isComplex()) {
    return caseField.hasNoHiddenSubFields(); // 1. This makes no sense to me. It seems the opposite of what we want.
  }
  return caseField.isHidden(); // 2. This bit makes sense.
  // 3. Returning at 1. for ALL complex fields also removes the possibility of the complex field
  //    ITSELF being hidden... is that not possible?
}
```

Non-complex fields:
1. CaseField is HIDDEN
** `isHidden(...) => true` ✅
2. CaseField is SHOWN
** `isHidden(...) => false` ✅

Complex fields (2 sub-fields, A and B):
1. CaseField is HIDDEN (A is SHOWN, B is SHOWN)
** `isHidden(...) => true` ✅
2. CaseField is HIDDEN (A is HIDDEN, B is SHOWN)
** `isHidden(...) => false` ❌
3. CaseField is HIDDEN (A is SHOWN, B is HIDDEN)
** `isHidden(...) => false` ❌
4. CaseField is HIDDEN (A is HIDDEN, B is HIDDEN)
** `isHidden(...) => false` ❌
5. CaseField is SHOWN (A is SHOWN, B is SHOWN)
** `isHidden(...) => true` ❌
6. CaseField is SHOWN (A is HIDDEN, B is SHOWN)
** `isHidden(...) => false` ✅
7. CaseField is SHOWN (A is SHOWN, B is HIDDEN)
** `isHidden(...) => false` ✅
8. CaseField is SHOWN (A is HIDDEN, B is HIDDEN)
** `isHidden(...) => false` ✅

I _think_ the logic should be something more like this:
```
public isHidden(caseField: CaseField, editForm: FormGroup): boolean {
  const formFields = editForm.getRawValue();
  if (caseField.field_type.complex_fields && caseField.field_type.complex_fields.length > 0) {
    // Check to see if ALL of the sub-fields are hidden.
    const allHidden = caseField.field_type.complex_fields.every((cf: CaseField) => this.isHidden(cf, editForm));

    // If they're ALL hidden, this complex field should be considered hidden too so return early.
    if (allHidden) {
      return true;
    }
  }

  // If we get here, either it's not a complex field or it IS a complex field and has at least one visible sub-field
  // so we need to see whether the complex field ITSELF is hidden.
  const condition = ShowCondition.getInstance(caseField.show_condition);
  return !condition.match(formFields.data);
}
```